torchsynth/module.py

```diff
@@ -227,7 +227,65 @@ def update_device(self, device=None):
             parameter.parameter_range.to(device)
 
 
-class ADSR(SynthModule):
+class ControlRateModule(SynthModule):
+    """
+    Control rate module is an abstract base class for SynthModules that run
+    at a control rate as opposed to full audio rate.
+
+    Optionally, the output control signal can be upsampled to the correct
+    sample rate and buffer size.
+    """
+
+    def __init__(
+        self,
+        synthglobals: SynthGlobals,
+        upsample: Optional[bool] = False,
+        **kwargs: Dict[str, T],
+    ):
+        super().__init__(synthglobals, **kwargs)
+        self.upsample = upsample
+        if upsample:
+            self.upsample = torch.nn.Upsample(
+                self.synthglobals.buffer_size, mode="linear", align_corners=True
+            )
+
+    @property
+    def sample_rate(self) -> T:
+        """
+        The internal sample rate for this module is control rate
+        """
+        assert self.synthglobals.control_rate.ndim == 0
+        return self.synthglobals.control_rate
+
+    @property
+    def buffer_size(self) -> T:
+        """
+        The internal buffer size for this module is the control buffer size
+        """
+        assert self.synthglobals.control_buffer_size.ndim == 0
+        return self.synthglobals.control_buffer_size
+
+    def forward(self, *args: Any, **kwargs: Any) -> Signal:  # pragma: no cover
+        """
+        Wrapper for _forward that ensures a buffer_size length output.
+        """
+        signal = self._forward(*args, **kwargs)
+        if self.upsample:
+            signal = self.upsample(signal.unsqueeze(1)).squeeze(1)
```

Comment:
Again, I still think control rate should factor nicely into the signal_rate so we don't need to fix lengths.

Reply:
Agreed. This fix length is just to be consistent with SynthModule since that runs fix_length on every output